pkg/local_object_storage/blobstor/peapod/peapod.go

```diff
@@ -0,0 +1,484 @@
+			return
+		case <-t.C:
+			st := time.Now()
+
+			x.flushCurrentBatch(true)
+
+			interval := flushInterval - time.Since(st)
+			if interval <= 0 {
+				interval = time.Microsecond
+			}
+
+			t.Reset(interval)
+		}
+	}
+}
+
+func (x *Peapod) flushCurrentBatch(beginNew bool) {
+	x.currentBatchMtx.Lock()
+
+	if !x.currentBatch.nonIdle {
+		if !beginNew && x.currentBatch.tx != nil {
+			_ = x.currentBatch.tx.Commit()
+		}
+		x.currentBatchMtx.Unlock()
+		return
+	}
+
+	err := x.currentBatch.tx.Commit()
+	if err != nil {
+		err = fmt.Errorf("commit BoltDB batch transaction: %w", err)
+	}
+
+	x.currentBatch.commitErr = err
+	close(x.currentBatch.chCommitted)
+
+	if beginNew {
+		x.beginNewBatch()
+	}
+
+	x.currentBatchMtx.Unlock()
+}
+
+func (x *Peapod) beginNewBatch() {
+	x.currentBatch = new(batch)
+
+	x.currentBatch.tx, x.currentBatch.initErr = x.bolt.Begin(true)
+	if x.currentBatch.initErr != nil {
+		x.currentBatch.initErr = fmt.Errorf("begin new BoltDB writable transaction: %w", x.currentBatch.initErr)
+		return
+	}
+
+	x.currentBatch.bktRoot = x.currentBatch.tx.Bucket(rootBucket)
+	if x.currentBatch.bktRoot == nil {
+		x.currentBatch.initErr = errMissingRootBucket
+		return
+	}
+
+	x.currentBatch.chCommitted = make(chan struct{})
+}
+
+const objectAddressKeySize = 2 * sha256.Size
+
+func keyForObject(addr oid.Address) []byte {
+	b := make([]byte, objectAddressKeySize)
+	addr.Container().Encode(b)
+	addr.Object().Encode(b[sha256.Size:])
+	return b
+}
+
+func decodeKeyForObject(addr *oid.Address, key []byte) error {
+	if len(key) != objectAddressKeySize {
+		return fmt.Errorf("invalid object address key size: %d instead of %d", len(key), objectAddressKeySize)
+	}
+
+	var cnr cid.ID
+	var obj oid.ID
+
+	err := cnr.Decode(key[:sha256.Size])
+	if err != nil {
+		return fmt.Errorf("decode container ID: %w", err)
+	}
+
+	err = obj.Decode(key[sha256.Size:])
+	if err != nil {
+		return fmt.Errorf("decode object ID: %w", err)
+	}
+
+	addr.SetContainer(cnr)
+	addr.SetObject(obj)
+
+	return nil
+}
+
+// Open opens underlying database in the specified mode.
+func (x *Peapod) Open(readOnly bool) error {
+	err := util.MkdirAllX(filepath.Dir(x.path), x.perm)
+	if err != nil {
+		return fmt.Errorf("create directory '%s' for database: %w", x.path, err)
+	}
+
+	x.bolt, err = bbolt.Open(x.path, x.perm, &bbolt.Options{
+		ReadOnly: readOnly,
+		Timeout:  100 * time.Millisecond, // to handle flock
+	})
+	if err != nil {
+		return fmt.Errorf("open BoltDB instance: %w", err)
+	}
+
+	if readOnly {
+		err = x.bolt.View(func(tx *bbolt.Tx) error {
+			if tx.Bucket(rootBucket) == nil {
+				return errMissingRootBucket
+			}
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("check root bucket presence in BoltDB instance: %w", err)
+		}
+	}
+
+	x.readOnly = readOnly
+
+	return nil
+}
+
+// Init initializes internal structure of the underlying database.
+func (x *Peapod) Init() error {
+	if x.readOnly {
+		// no extra actions needed in read-only mode
```

Comment:
Why don't you just check for the root bucket? It's a kind of a sanity check for the DB in our case.

Reply:
checked in `Open`